buckwheat/tokenizer.py

```diff
@@ -140,11 +144,18 @@ def traverse_tree(node: tree_sitter.Node, types: Set[str]) -> Iterator[tree_sitt
         :param types: the set of types of interest.
         :return: the iterator of Tree-sitter nodes of necessary types.
         """
-        for child in node.children:
-            if child.type in types:
-                yield child
-            if len(child.children) != 0:
-                yield from TreeSitterParser.traverse_tree(child, types)
+        current_depth = 0
+        stack = deque([(node, current_depth)])
+
+        while stack:
+            if current_depth > MAX_AST_DEPTH:
+                raise DeepASTError("Current AST traversal is too deep")
```

Comment:
why do you want to throw `DeepASTError`?

Reply:
Because as @areyde said, files with deep ASTs probably aren't useful for researchers, because they always contain something strange like configs/tests/etc. 

Also it was the behaviour of tool when traverse was made by recursion and I surely don't want to break that for now, because there are some tools using it.

Reply:
`something strange like configs/tests/etc` - it will be good to provide examples. TBH it's strange that non-recursive implementation has some depth limits.
`traverse was made by recursion and I surely don't want to break that for now` - it was done because it was fast implementation - so I don't see the reason to keep recursive implementation limitations

Reply:
He's talking about our concerns. I personally have never seen a file with a 1000 levels of AST that would be completely _normal_. And since we always gathered identifiers on the level of projects, we were fine with skipping such files. You may not keep the limitations, sure, but I suggested expermenting with depth and finding some heuristic that would be possible to turn on — to skip _anomalous_ files.

Reply:
I agree with @EgorBu that depth limit here is very strange and not obvious. I think we could remove this recursive limitation from here and just add possibility to filter out files with deep ASTs in later processing.  It could be done easily by extending `FileData` class with ast_depth parameter.

Reply:
Interesting idea!

Reply:
Either that, or ignore it.

Reply:
For this PR I think we should ignore it and left everything as is. IMO, deep ASTs filtration should be done as distinct feature and have separate PR.